ethers-core/src/types/transaction/eip712.rs

```diff
@@ -0,0 +1,276 @@
+//! TL;DR you're probably looking for `ethers-derive-eip712` Eip712 derive macro.
+//!
+//! The eip712 module contains helper methods and types mainly used
+//! by the derive-eip712 procedural macro. Note that many of the methods
+//! used in this module may panic!. While this is desired behavior for a
+//! procedural macro, it may not be the behavior you wish to use in your
+//! application if using these methods manually.
+use std::collections::HashMap;
+
+use convert_case::{Case, Casing};
+
+use crate::{
+    abi,
+    abi::Token,
+    types::{Address, H160, U256},
+    utils::keccak256,
+};
+
+/// Error typed used by Eip712 derive macro
+#[derive(Debug, thiserror::Error)]
+pub enum Eip712Error {
+    #[error("Failed to serialize serde JSON object")]
+    SerdeJsonError(#[from] serde_json::Error),
+    #[error("Failed to decode hex value")]
+    FromHexError(#[from] hex::FromHexError),
+    #[error("Failed to make struct hash from values")]
+    FailedToEncodeStruct,
+}
+
+/// The Eip712 trait provides helper methods for computing
+/// the typed data hash used in `eth_signTypedData`.
+///
+/// The ethers-rs `derive_eip712` crate provides a derive macro to
+/// implement the trait for a given struct. See documentation
+/// for `derive_eip712` for more information and example usage.
+///
+/// For those who wish to manually implement this trait, see:
+/// https://eips.ethereum.org/EIPS/eip-712
+///
+/// Any rust struct implementing Eip712 must also have a corresponding
+/// struct in the verifying ethereum contract that matches its signature.
+///
+/// NOTE: Due to limitations of the derive macro not supporting return types of
+/// [u8; 32] or Vec<u8>, all methods should return the hex encoded values of the keccak256
+/// byte array.
+pub trait Eip712 {
+    /// User defined error type;
+    type Error: std::error::Error + Send + Sync + std::fmt::Debug;
+
+    /// The eip712 domain is the same for all Eip712 implementations,
+    /// This method does not need to be manually implemented, but may be overridden
+    /// if needed.
+    fn eip712_domain_type_hash() -> String {
+        hex::encode(crate::utils::keccak256(
+            "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)",
+        ))
+    }
+
+    /// The domain separator depends on the contract and unique domain
+    /// for which the user is targeting. In the derive macro, these attributes
+    /// are passed in as arguments to the macro. When manually deriving, the user
+    /// will need to know the name of the domain, version of the contract, chain ID of
+    /// where the contract lives and the address of the verifying contract.
+    fn domain_separator() -> String;
+
+    /// This method is used for calculating the hash of the type signature of the
+    /// struct. The field types of the struct must map to primitive
+    /// ethereum types or custom types defined in the contract.
+    fn type_hash() -> String;
```

Comment:
The original error I was getting was due to trying to pass the compile time [u8; 32] into the method using #type_hash. However, this does not work due to ToTokens not being implemented for [u8; 32].

That said, I did manage to update the trait to return `Result<[u8; 32], Eip712Error>` rather than `String` simply by performing the hex encoding -> String at compile time and decoding the `'static &str` -> `[u8;32]` within the methods. It's a hack, but it allows us to return the desired `[u8; 32]` type, avoiding the user having to worry about hex encoding/decoding.

Reply:
I'm also not opposed to returning Bytes.

Reply:
decided against using bytes due to conversion overhead, using `[u8;32` return type.